File: session04/primes.cc
#include <iostream>
using namespace std;

// true means n is prime
//O(n)  omega(1)
bool bruteForceIsPrime(uint64_t n) {   // n =2
  for (uint64_t i = 2; i < n; i++) {
    if (n % i == 0)
      return false;
  }
  return true;
}

/*
  28

2*14
4*7


 */

//O(sqrt(n))
bool bruteForceIsPrime2(uint64_t n) { 
  for (uint64_t i = 2; i <= sqrt(n); i++) {
    if (n % i == 0)
      return false;
  }
  return true;
}

//O(n*sqrt(n))
uint64_t countPrimes(uint64_t n) {
  uint64_t count = 0;
  for (uint64_t i = 2; i <= n; i++)
    if (bruteForceIsPrime(i))
      count++;
  return count;
}

int main() {
  cout << bruteForceIsPrime(100000000000000001) << '\n';
  
  cout << countPrimes(n) << '\n';

}

File: old/session04_Primes/Primes.java

public class Primes {
    public boolean bruteForceIsPrime(long n) {
	for (long i = 2; i < n; i++) //O(n)
	    if (n % i == 0)
		return false; // omega(1)
	return true; //O(n)
    }

    // n=10^12 sqrt(n) = 10^6
    public boolean bruteForceIsPrime2(long n) { // case 9
	for (long i = 2; i <= Math.sqrt(n); i++) //O(sqrt(n))
	    if (n % i == 0)
		return false; // omega(1)
	return true; //O(n)
    }

    //O(n*sqrt(n))  n=10^9 sqrt(n) = 10^4.5 == 33000
    public int countPrimes(long n) {
	int count = 0;
	for (long i = 2; i <= n; i++)  //O(n)
	    if (bruteForceIsPrime2(i)) // O(sqrt(n))
		count++;
	return count;
    }

        // n=10^12 sqrt(n) = 10^6
    public boolean bruteForceButTrickier(long n) { // case 9
	for (long i = 3; i <= Math.sqrt(n); i += 2) //O(sqrt(n))
	    if (n % i == 0)
		return false; // omega(1)
	return true; //O(n)
    }

    public int countPrimes2(long n) {
	int count = 1; // 2 is prime, special case
	for (long i = 3; i <= n; i += 2)  //O(n)
	    if (bruteForceIsPrime2(i)) // O(sqrt(n))
		count++;
	return count;
    }

    
    public static void main(String[]args) {
	//	System.out.println(bruteForceIsPrime(10000000000000001));
	System.out.println(bruteForceIsPrime2(10000000000000001));
    }

}
